lib/common/xml.c

```diff
@@ -1536,39 +1551,40 @@ log_xml_node_and_children(GString *buffer, int log_level, const char *prefix,
         return;
     }
 
-    name = crm_element_name(data);
-
-    g_string_truncate(buffer, 0);
-
     if (pcmk_is_set(options, xml_log_option_open)) {
+        g_string_truncate(buffer, 0);
         insert_prefix(options, buffer, depth);
 
         if (data->type == XML_COMMENT_NODE) {
             pcmk__g_strcat(buffer, "<!--", (const char *) data->content, "-->",
                            NULL);
 
-        } else {
-            pcmk__g_strcat(buffer, "<", name, NULL);
+        } else {    // data->type == XML_ELEMENT_NODE
+            const char *hidden = crm_element_value(data, "hidden");
 
-            hidden = crm_element_value(data, "hidden");
-            for (xmlAttrPtr a = pcmk__xe_first_attr(data); a != NULL;
-                 a = a->next) {
+            pcmk__g_strcat(buffer, "<", name, NULL);
 
-                xml_node_private_t *nodepriv = a->_private;
-                const char *p_name = (const char *) a->name;
-                const char *p_value = pcmk__xml_attr_value(a);
+            for (const xmlAttr *attr = pcmk__xe_first_attr(data); attr != NULL;
+                 attr = attr->next) {
+                xml_node_private_t *nodepriv = attr->_private;
+                const char *p_name = (const char *) attr->name;
+                const char *p_value = pcmk__xml_attr_value(attr);
                 char *p_copy = NULL;
 
                 if (pcmk_is_set(nodepriv->flags, pcmk__xf_deleted)) {
                     continue;
-                } else if (pcmk_any_flags_set(options,
-                                              xml_log_option_diff_plus
-                                              |xml_log_option_diff_minus)
-                           && (strcmp(XML_DIFF_MARKER, p_name) == 0)) {
+                }
+
+                if (pcmk_any_flags_set(options,
+                                       xml_log_option_diff_plus
+                                       |xml_log_option_diff_minus)
+                    && (strcmp(XML_DIFF_MARKER, p_name) == 0)) {
                     continue;
+                }
 
-                } else if (hidden != NULL && p_name[0] != 0 && strstr(hidden, p_name) != NULL) {
-                    p_copy = strdup("*****");
+                if ((hidden != NULL) && (p_name[0] != '\0')
+                    && (strstr(hidden, p_name) != NULL)) {
```

Comment:
We can modify it (either now or later) to use a regex comparison or something, if we define how the hidden list will be delimited.

Currently, `"hidden"` is used in only one place, and it contains only one item.
```
lib/cib/cib_remote.c:
static int
cib_tls_signon(cib_t *cib, pcmk__remote_t *connection, gboolean event_channel)
{
...
    crm_xml_add(login, "hidden", "password");
...
}
```

Maybe we're planning to extend it to allow other hidden attributes? I don't see a way to do so yet without an additional patch. If not, then it would be simpler to compare against `"password"` instead of searching through a `"hidden"` list.

Reply:
It might be nice to be able to define a list of hidden attributes in general though.

Reply:
Looks like this wasn't particularly thought through, just a one-off for what was needed.  It would have been better to make this a sub-element like `<sensitive><attr name="password"/></sensitive>` but that wouldn't be backward compatible now. I would just ignore this for now since we don't have an immediate use for improving it. I'll add a task.